src/item.cpp

```diff
@@ -6393,186 +6388,161 @@ void item::update_temp( const int temp, const float insulation )
+    float true_energy = mass * thermal_energy * 0.001;
+    
+    // If item has not been processed for two days just set it to environment temperature
+    // Or if the item has negative energy (has not been processed ever)
+    // This can cause incorrect freezing/unfreezing when the temperature is close to freezing temperature and the freezing/unfreezing would take longer than two days
+    if ( to_turns<int>( time ) > 28800 || thermal_energy < 0 ){
+        add_msg( _( "New or very old item" ));
+        true_energy = 1000 * get_energy_from_temperature( env_temperature ) / mass + 0.5;
+        temperature = static_cast<int>( 10 * env_temperature );
+        thermal_energy = static_cast<int>( true_energy );
+        return;
+    }
+    
+    // thermal_energy = item thermal energy (mJ/g). Stored in the item
+    // temperature = item temperature (dK deci (=0.1) kelvins). Stored in the item
+    const float conductivity_term = 20 / insulation; // Constant chosen to "feel good". Insulation is also applied here
+    const float freezing_temperature = ( type->comestible->freeze_point - 32 ) * 0.556 + 273.15;  // F converted to K
+    const int specific_heat_liquid = type->comestible->spec_heat_liquid; // J/g K
+    const int specific_heat_solid = type->comestible->spec_heat_solid; // J/g K
+    const int latent_heat = type->comestible->lat_heat; // J/kg
+    const float surface_area = pow( 0.000001 * to_milliliter( volume() ), 0.66 ); // Volume (m3)^2/3
+    const float temperature_difference =  env_temperature - old_temperature;
+    float freeze_percentage = 0;
+    
+    float energy_change = temperature_difference * conductivity_term * surface_area * 6 *to_turns<int>( time );
+    true_energy += energy_change;
+    const float completely_frozen_energy = mass * specific_heat_solid * freezing_temperature;  // Energy that the item would have if it was completely solid at freezing temperature
+    const float completely_liquid_energy = completely_frozen_energy + mass * latent_heat; // Energy that the item would have if it was completely liquid at freezing temperature    
+    float new_item_temperature;
+    
+    if( true_energy > completely_liquid_energy ) {
+        // Item is liquid
+        new_item_temperature = freezing_temperature + ( true_energy - completely_liquid_energy ) / ( specific_heat_liquid * mass );
+    } else if( true_energy < completely_frozen_energy ) {
+        // Item is solid
+        freeze_percentage = 1;
+        new_item_temperature =  true_energy / ( specific_heat_solid * mass );
+    } else {
+        // Item is partially solid
+        new_item_temperature = freezing_temperature;
+        freeze_percentage = ( completely_liquid_energy - true_energy ) / ( completely_liquid_energy - completely_frozen_energy );
+    }
+    // Stop over cooling below environment
+    // Stop over heating above environment
+    if( energy_change < 0 && new_item_temperature < env_temperature ) {
```

Comment:
Is this part really required? It will never be used if everything working nicely. And it will not solve a problem if energy change is much higher than it should be, it will only make this bug less obvious.
I suggest to remove it.

Reply:
It will be used when you return to an item that was outside of your reality bubble as the calulations will be done with large value for time.

Reply:
This error is from solving the differential equation by using linear approximation instead of correct solution (no idea what does it means, but it sounds very scientific). This approximation is acceptable only for short time periods and it gives very large error if used over long time.
It can be resolved by utilizing new variable similar to ``time_point last_rot_check`` and if the time since last recalculation is more then 6 hours - skip calculation and set item temperature equal to the air temperature. Using ``last_temp_check`` to measure time between updates is even better, because ``counters`` are changed only for items in the bubble and will not be updated if the item is outside.

Reply:
Yeah this approximation only holds when the time interval is short or the temperature difference is small.
And after checking we seem to be already in area where this approximation starts to have issues.
(I did all the tests with item somewhat close to enviroment temperature. But cooking food in ice lab or during winter can cause the item to cool down to enviroment temp instantly)

The correct solution is solving the equation as first order differential equation (=write it in differential form in wolfram alpha).
The proper equation (when item is solid from start to finish) is:

Energy at time t = E_0 * E_e * exp( - t * A * I / (c * m) ) + E_e

where
E_0 = energy at time = 0
E_e = Energy at item at enviroment temperature
t = Time
A = Surface area
I = Our conductivity constant
c = Specific hat for solid
m = mass

(different similar equations for liquid and partially solid)

This equation would give same correct solution no matter how long time it has been since previous calculation.

It is quite a bit more complex but I quess it has to be used.

Reply:
> But cooking food in ice lab or during winter can cause the item to cool down to enviroment temp instantly


And that is why linear approximation is not the perfect one. But it still can be used in this situation if the energy will be recalculated on every turn.

>It is quite a bit more complex but I quess it has to be used.


Just replace 
``float energy_change = temperature_difference * conductivity_term * surface_area * 6 *to_turns<int>( time );``
with this one
``float energy_change = ( get_energy_from_temperature( env_temperature )  - true_energy ) * exp(  -1 * abs( temperature_difference ) * conductivity_term * surface_area * 6 * to_turns<int>( time ) );``

And you should change ``conductivity_term`` to make last part dimensionless.

Protip: never take codes from a random guy on the internet.

Trigonometric, exponential and other suspicious functions are rather expensive in terms of calculation time, because they are expanded into near infinite Taylor series. And using them everywhere is not the best, that is why other numerical solutions were created. So avoid them if the calculation time is the main problem.

Reply:
My previous equation was wrong. Asked on some anime forums and they told me what the temperature difference in the initial heat flow rate equation can be replaced with the energy difference, because temperature is a function of the entropy, i.e. internal energy. So in the differential form this equation will be:

dQ/dt = ( Q_e - Q_0 ) * I * A / m

where Q  - energy, Q_0 - initial energy, Q_e - energy at item at environment temperature, t  - time, A - surface area, I - our *updated*  conductivity constant, m - mass.
And the solution Q_t ( energy after time t ) will be:

Q_t = Q_0 + ( Q_e - Q_0 ) * ( 1 - exp( -t *  I * A / m ) )

No proof here. It is magic. Energy difference for one time step is:

dQ = Q_t - Q_0 = ( Q_e - Q_0 ) * ( 1 - exp( -t *  I * A / m ) )

This gives us final ingame text:

``float energy_change = ( get_energy_from_temperature( env_temperature )  - true_energy ) * ( 1 - exp(  conductivity_term * surface_area / mass * 6 * to_turns<int>( time ) ) );``

Sidenotes:
6 should be removed. Either make conductivity_term per turn instead of per second, or use game constant. It must be something like seconds_in_one_turn.
Value of conductivity_term should be changed, because it is related to the mass now.

Reply:
I ended up with similar but slightly different equation:

T_t= ( T_0-  T_e) * exp( - time * surface_area * conductivity_term / ( mass * specific_heat) ) + T_e

And then energy is calculated from the temperature. Odd that you don't have specific heat as that should affect how much the item heats up (smaller specific heat -> bigger temperature change -> smaller temperature difference -> less energy flow).

We don't really need the energy change anymore at all. We get the temperature and energy at time t directly.

Putting the 6 in the conductivity term is good idea (simply multiplied it with 6).

I did a test with the performance. I collecteda 347 m3 of food and put them in a vehicle.
In current game waiting 8 hours takes ~9 seconds
With new temperature system 8 hours takes ~11 seconds.

Maybe the linear approximation could be used when the temperature difference is small. I should have tested the performance before switching to the new formula so I would know if the exp is even causing any noticeable difference.

Reply:
Since the difference in performance from temperature calculations is so small that I can't even measure it by hand I think the current math is good enough.

I have one more fix on the math under work that will fix all issues with solid-melting-liquid transitions.